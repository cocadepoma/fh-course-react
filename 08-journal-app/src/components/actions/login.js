import { types } from "../../types/types";

export const setLoginError = (err) => ({
    type: types.loginSetError,
    payload: err,
});

export const removeLoginError = () => ({
    type: types.loginRemoveError,
});

export const startLoading = () => ({
    type: types.loginStartLoading,
});

export const finishLoading = () => ({
    type: types.loginFinishLoading,
});
