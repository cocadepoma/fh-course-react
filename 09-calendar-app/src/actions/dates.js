import { types } from "../types/types";

export const setDate = (date) => ({
    type: types.dateSet,
    payload: date,
});
export const clearDate = () => ({
    type: types.dateClear,
});
