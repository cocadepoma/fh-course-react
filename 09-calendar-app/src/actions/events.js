import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";


export const eventStartAddNew = (event) => {
    return async (dispatch, getState) => {

        const { name, uid } = getState().auth;

        try {
            const resp = await fetchConToken('events', event, 'POST');
            const body = await resp.json();

            if (body.ok) {

                event.id = body.event.id;
                event.user = {
                    name: name,
                    _id: uid
                }

                dispatch(eventAddNew(event));
            }
            console.log(body)
        } catch (error) {
            console.log(error);
        }
    }
}

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event,
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event,
});

export const eventClearActive = () => ({
    type: types.eventClearActive,
});

export const eventStartUpdate = (event) => {
    return async (dispatch) => {
        try {

            const resp = await fetchConToken(`events/${event.id}`, event, 'PUT');
            const body = await resp.json();

            if (body.ok) {
                dispatch(eventUpdate(event));
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        }
    }
};


export const eventUpdate = (event) => ({
    type: types.eventUpdate,
    payload: event,
});

export const eventStartDeleting = () => {

    return async (dispatch, getState) => {

        const { activeEvent } = getState().calendar;
        const resp = await fetchConToken(`events/${activeEvent.id}`, {}, 'DELETE');
        const body = await resp.json();

        if (body.ok) {
            dispatch(eventDelete());
        } else {
            Swal.fire('Error', body.msg, 'error');
        }

    }
}

const eventDelete = () => ({
    type: types.eventDelete,
});

export const eventStartLoading = () => {
    return async (dispatch) => {

        try {

            const resp = await fetchConToken('events');
            const body = await resp.json();
            const events = prepareEvents(body.events);


            if (body.ok) {
                dispatch(eventLoaded(events));
            } else {
                Swal.fire('Error', 'Error al obtener los eventos', 'error');
            }

        } catch (error) {
            console.log(error);
        }
        console.log('>?>>>')
    }
}

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventCleaner = () => ({
    type: types.eventCleaner
});

