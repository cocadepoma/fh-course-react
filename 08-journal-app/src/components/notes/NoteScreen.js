import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "../../hooks/useForm";
import { activeNote, startDeleting } from "../actions/notes";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
    const dispatch = useDispatch();
    const { active: note } = useSelector((state) => state.notes);
    const [formValues, handleInputChange, reset] = useForm(note);
    const { body, title, id } = formValues;

    const activeId = useRef(note.id);

    // Para detectar cuando refrescar la nota, nos creamos una referencia
    // Por ejemplo de su id y la guardarmos en un useRef. Después en el useEffect si detectamos
    // que son diferentes, las establecemos y reseteamos el form con la nueva nota y cambiamos el useRef
    // Como dependencias del useEffect establecemos la nota y el reset, así cada vez que
    // cambie la nota comprobará si son distintos los ID's y en caso de serlo, hará el reset y volverá
    // a cargar la data.
    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }));
    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch(startDeleting(id));
    };

    return (
        <div className='notes__main-content animate__animated animate__fadeInRightBig animate__faster'>
            <NotesAppBar />
            <div className='notes__content'>
                <input
                    type='text'
                    placeholder='Some awesome title'
                    className='notes__title-input'
                    autoComplete='off'
                    value={title}
                    onChange={handleInputChange}
                    name='title'
                />
                <textarea
                    className='notes__textarea'
                    placeholder='What happened today'
                    value={body}
                    onChange={handleInputChange}
                    name='body'
                ></textarea>

                {note.url && (
                    <div className='notes__image'>
                        <img src={note.url} alt='landscape' />
                    </div>
                )}
            </div>

            <button className='btn btn-danger' onClick={handleDelete}>
                Delete
            </button>
        </div>
    );
};
