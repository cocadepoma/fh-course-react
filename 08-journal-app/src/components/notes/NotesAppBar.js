import React from "react";
import moment from "moment";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../actions/notes";

export const NotesAppBar = ({ id, date }) => {
    const formatDate = moment(date);

    const dispatch = useDispatch();
    const { active: note } = useSelector((state) => state.notes);

    const handleSave = () => {
        dispatch(startSaveNote(note));
    };
    const handlePictureUpload = () => {
        document.querySelector("#fileSelector").click();
    };

    const handleFileChange = (e) => {
        const file = e.target.files[0];

        if (file) {
            dispatch(startUploading(file));
        }
    };
    return (
        <div className='notes__appbar'>
            <span>{formatDate.format("LL")}</span>
            <input
                id='fileSelector'
                type='file'
                style={{ display: "none" }}
                onChange={handleFileChange}
                name='file'
            />
            <div>
                <button className='btn' onClick={handlePictureUpload}>
                    Picture
                </button>
                <button className='btn' onClick={handleSave}>
                    Save
                </button>
            </div>
        </div>
    );
};
