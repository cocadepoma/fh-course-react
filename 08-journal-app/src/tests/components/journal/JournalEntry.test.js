import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";

import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../components/actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const note = {
    id: 123,
    title: "hola",
    date: 0,
    body: "Mundo",
    url: "https://github.com",
};

const wrapper = mount(
    <Provider store={store}>
        <JournalEntry {...note} />
    </Provider>
);

describe("pruebas en <JournalEntry/>", () => {
    test("debe de mostrarse correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("debe de poner la nota en active", () => {
        wrapper.find(".journal__entry").prop("onClick")();

        // En lugar de hacer un mock, también se puede comprobar con que parámetros ha sido llamada
        // Hay que tener en cuenta que siempre que tengamos diferentes dispatch puede llevarnos a confusión
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, { ...note }));
    });
});
