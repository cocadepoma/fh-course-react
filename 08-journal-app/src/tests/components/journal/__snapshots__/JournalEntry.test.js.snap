// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <JournalEntry/> debe de mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__fast"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://github.com)",
        "backgroundPosition": "center center",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body-date"
  >
    <div
      className="journal__entry-body"
    >
      <p
        className="journal__entry-title"
      >
        hola
      </p>
      <p
        className="journal__entry-content"
      >
        Mundo
      </p>
    </div>
    <div
      className="journal__entry-date-box"
    >
      <span>
        Thursday
      </span>
      <h4>
        1st
      </h4>
    </div>
  </div>
</div>
`;
